File: codility/lessons/sorting/MaxProductOfThree.scala
package codility.lessons.sorting

import scala.util.Random

/**
 * Painless
 */
object MaxProductOfThreeScala extends App {
    object Solution {
        def solution(arr: Array[Int]) = { 
            val sorted = arr.sorted
            val n      = sorted.length 
            Math.max(sorted(0    ) * sorted(1    ) * sorted(n - 1), 
                     sorted(n - 3) * sorted(n - 2) * sorted(n - 1))
         }
    }
}

File: codility/lessons/sorting/MaxProductOfThree.java
package codility.lessons.sorting;

import java.util.*;

/**
 * Level : painless
 */
class MaxProductOfThree {
    public int solution(int[] A) {
        int[] B = Arrays.copyOf(A,A.length);
        Arrays.sort(B);
        int max = B[0]*B[1]*B[B.length - 1];
        if (max > 0) 
            return Math.max(max,B[B.length - 3]*B[B.length - 2]*B[B.length - 1]);
        for (int i=2 ; i<B.length ; max = Math.max(max,B[i]*B[i-1]*B[i-2]), i++);
        return max;
    }
}


